e2e/playground.py

```diff
@@ -0,0 +1,2135 @@
+            # # Simple approach if position wasn't needed
+            # has_locator = self.page.locator(
+            #     f"{el_type}[{key}='{item}']{is_checked_str}"
+            # )
+
+            has_locator = self.page.locator(f"{el_type}{is_checked_str}")
+            # Go up one element as CSS does not have "self" selector
+            # We can leverage this as the HTML structure is very nested.
+            # This approach will break if there are matching siblings
+            has_locator = has_locator.nth(i).locator("..")
+            # Find direct child that matches selector
+            has_locator = has_locator.locator(
+                f"> {el_type}[{key}='{item}']{is_checked_str}"
+            )
+
+            loc_container = loc_container.locator(
+                # Return self
+                "xpath=.",
+                has=has_locator,
+            )
+        # Make sure other items are not in set
+        loc_inputs = loc_container.locator(f"{el_type}{is_checked_str}")
+        # TODO-barret; Look into adding a try-catch around this and then performing the locator check using multiple expectations to get better error messages
+        playwright_expect(loc_inputs).to_have_count(len(arr), timeout=timeout)
+
+
+class InputCheckboxGroup(
+    _WidthContainer,
+    _RadioButtonCheckboxGroupBase,
+):
+    # label: TagChildArg,
+    # choices: ChoicesArg,
+    # selected: Optional[Union[str, List[str]]] = None,
+    # inline: bool = False,
+    # width: Optional[str] = None,
+    def __init__(
+        self,
+        page: Page,
+        id: str,
+    ) -> None:
+        super().__init__(
+            page,
+            id=id,
+            loc="label input[type=checkbox]",
+            loc_container=f"div#{id}.shiny-input-checkboxgroup.shiny-bound-input",
+            loc_label=f"label#{id}-label",
+        )
+
+        # `loc_container` does not need to contain checked items
+        # But we should have `self.loc` be `self.loc_selected`
+        self.loc = self.loc_container.locator("label input[type=checkbox]:checked")
+        # Same value
+        self.loc_selected = self.loc
+
+        self.loc_choices = self.loc_container.locator("label input[type=checkbox]")
+        self.loc_choice_labels = self.loc_container.locator(
+            "label",
+            has=self.page.locator("input[type=checkbox]"),
+        )
+
+    def set(
+        self,
+        selected: typing.Union[str, typing.List[str]],
+        *,
+        timeout: Timeout = None,
+        **kwargs: typing.Any,
+    ) -> None:
+
+        if isinstance(selected, str):
+            selected = [selected]
+        # We are wanting to delay retriving the value of the checkbox as long as possible
+        checkbox_loc = self.loc_choices
+        checkbox_loc.nth(0).wait_for(state="attached", timeout=timeout)
+
+        # TODO-barret; Could do with multiple locator calls, but unchecking the values without the known values would be difficult.
```

Comment:
It would be great to have a variant of `expect_choices()` to be called here, that ensures that all the choices you want to set/unset are actually present.

Reply:
e712415